### PR TITLE
Calculate per-sec network metrics for RXBytes and TxBytes

### PR DESCRIPTION
**Pull request by ubhattacharjya**
### Summary

Calculate RxBytes/sec and TxBytes/sec metrics for bridge network mode.

### Implementation details

Added the fields RxBytesPerSecond and TxBytesPerSecond to the existing networkStats struct. The above mentioned network metrics for last stat in the queue will be stored in memory and we will find the difference in value of currentStats and lastStats and divide it with the timeDuration between the 2 stats in seconds.

### Testing

Started a task which creates network traffic and saw that the calculation was happening as expected.
Added unit test for the same.

New tests cover the changes: 

### Licensing

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: agent/stats/queue.go [hunk 1]
```diff
@@ -103,6 +103,13 @@ func (queue *Queue) add(rawStat *ContainerStats) {
 			stat.CPUUsagePerc = 100 * cpuUsageSinceLastStat / timeSinceLastStat
 		}
 
+		//calculate per second Network metrics
+		timeSinceLastStatInSeconds := uint64(rawStat.timestamp.Sub(lastStat.Timestamp).Seconds())
```
Comment (sparrc) [inline agent/stats/queue.go:107]: this is going to be a problem for streaming stats, which sometimes might return stats that are less than 1 second apart from each other. I think it would be better to determine the bytes per millisecond and then divide by 1000 to get the per-second average.

Comment (sparrc) [reply agent/stats/queue.go:107]: this would also mean you dont have to change the timestamps of the unit tests (which i think should stay as they are, since we need to support metrics arriving at less than 1-second intervals).

Comment (shubham2892) [reply agent/stats/queue.go:107]: > this is going to be a problem for streaming stats

For AWSVPC mode network stats we are only polling stats.

Comment (shubham2892) [reply agent/stats/queue.go:107]: Yes it makes sense to reuse `timeSinceLastStat` on line 96.

File: agent/stats/queue.go [hunk 2]
```diff
@@ -103,6 +103,13 @@ func (queue *Queue) add(rawStat *ContainerStats) {
...
+		if stat.NetworkStats != nil && timeSinceLastStatInSeconds > 0 {
+			stat.NetworkStats.RxBytesPerSecond = uint64(((stat.NetworkStats.RxBytes - lastStat.NetworkStats.RxBytes) / timeSinceLastStatInSeconds))
+			stat.NetworkStats.TxBytesPerSecond = uint64(((stat.NetworkStats.TxBytes - lastStat.NetworkStats.TxBytes) / timeSinceLastStatInSeconds))
```
Comment (shubham2892) [inline agent/stats/queue.go:110]: Will be sent for only bridge network mode? If not, we should add a check to ensure that this is only for bridge network mode task.

File: agent/stats/queue_test.go [hunk 3]
```diff
@@ -314,6 +314,18 @@ func validateNetStatsSet(t *testing.T, netStats *ecstcs.NetworkStatsSet, queueLe
 	assert.Equal(t, int64(queueLen), *netStats.TxErrors.SampleCount, "incorrect TxErrors sampleCount")
 	assert.Equal(t, int64(0), *netStats.TxErrors.Sum, "incorrect TxErrors sum")
 	assert.Equal(t, int64(0), *netStats.TxErrors.OverflowSum, "incorrect TxErrors overlfowSum")
+
+	assert.NotNil(t, *netStats.RxBytesPerSecond, "incorrect RxBytesPerSecond set")
+	assert.NotEqual(t, math.MaxFloat64, *netStats.RxBytesPerSecond.Min, "incorrect RxBytesPerSecond min")
```
Comment (sparrc) [inline agent/stats/queue_test.go:319]: why are these tests not asserting the actual value that we expect for RxBytesPerSecond and TxBytesPerSecond?

**Comment by sparrc** [comment]
There are some corner cases based on the fact that this metric depends on two different sets of stats from docker. For CPU percentage we have additional unit tests defined here: https://github.com/aws/amazon-ecs-agent/blob/master/agent/stats/queue_test.go#L395-L512 to cover some of these cases, I think we should have similar coverage for these metrics as well, such as:

1. what happens to these stats when there's only been 2 metrics from docker?
2. what happens to all of the network stats when there's only been 2 metrics from docker?
3. what happens to these when two metrics have the same timestamp?

File: agent/stats/queue.go (hunk 1 above)
Comment (ubhattacharjya) [reply agent/stats/queue.go]: Fixed it. Changed the unit of the per sec metrics to float32 which will help us to calculate the metric for even a small time duration

File: agent/stats/queue.go (hunk 2 above)
Comment (ubhattacharjya) [reply agent/stats/queue.go]: That check is being done here already https://github.com/aws/amazon-ecs-agent/blob/master/agent/stats/engine.go#L616 . That is why I am not doing it here

File: agent/stats/queue_test.go (hunk 3 above)
Comment (ubhattacharjya) [reply agent/stats/queue_test.go:319]: The actual float values are quite messy to assert. That is why just made sure that they are not empty and not not nil.

**Comment by ubhattacharjya** [comment]
> There are some corner cases based on the fact that this metric depends on two different sets of stats from docker. For CPU percentage we have additional unit tests defined here: https://github.com/aws/amazon-ecs-agent/blob/master/agent/stats/queue_test.go#L395-L512 to cover some of these cases, I think we should have similar coverage for these metrics as well, such as:
> 
> 1. what happens to these stats when there's only been 2 metrics from docker?
> 2. what happens to all of the network stats when there's only been 2 metrics from docker?
> 3. what happens to these when two metrics have the same timestamp?

This is a really good point. Added unit tests for each scenario.
1. If there are 2 metrics that should be fine. It needs atleast 2 metrics.
2. If there are 2 metrics the entire network stats metric should be fine. It needs atleast 2 metrics.
3. If there are 2 metrics only and they are both at the same timestamp, then it is a problem and it should throw error. But if there are more than 2 metrics and among them only 2 are same, it should be fine.
4. If there are less than 2 metrics, then it should throw error

File: agent/stats/queue_test.go (hunk 3 above)
Comment (sparrc) [reply agent/stats/queue_test.go:319]: you can use assert.InDelta to assert float values are within some delta of each other: https://pkg.go.dev/github.com/stretchr/testify/assert?tab=doc#InDelta

Comment (ubhattacharjya) [reply agent/stats/queue_test.go]: Fixed it
